compiler/rustc_middle/src/mir/consts.rs

```diff
@@ -318,8 +319,14 @@ impl<'tcx> Const<'tcx> {
             Const::Ty(_, c) => {
                 // We want to consistently have a "clean" value for type system constants (i.e., no
                 // data hidden in the padding), so we always go through a valtree here.
-                let (ty, val) = c.eval(tcx, param_env, span)?;
-                Ok(tcx.valtree_to_const_val((ty, val)))
+                match c.eval_valtree(tcx, param_env, span) {
```

Comment:
Hm, I quite specifically avoided that when writing this code -- things were arranged to be nicely symmetric. It almost seems to me like `eval_valtree` should now be renamed to `eval`?

Reply:
I would expect `eval_valtree` to not be around once all this refactoring is done, for the same reason as all the other `eval_*`. I think we should probably be using normalize erasing regions here instead? It's a bit unfortunate that this loses the information about *why* normalization failed though it seems like we only really use it for ICEs so that's probably fine

Reply:
That sounds very odd to me. The fundamental thing we can do with a (fully monomorphized) type system constant is get its valtree. A bunch of other systems need that, e.g. the interpreter. Why should we not have such an operation and how else should other parts of the compiler get the valtree?

Reply:
There is still ways to normalize a constant its not like its impossible- its just the the API is different and you would have to do the same thing you would do with a fully monomorphized type if you wanted to match on it to see what concrete type it is. It would look something like `normalize_erasing_regions(c).to_valtree()`.

Reply:
We're removing the ways of normalizing type system constants that dont work in all situations as that would be a footgun and potentially lead to soundness bugs. We're *not* making it impossible to convert an unevaluated constant into an evaluated constant and inspect the value

Reply:
I guess put another way: evaluation is a subset of type system normalization. Having evaluation methods available on `ty::Const` is a footgun as people will use them and think they perform normalization (normalization is definitely one of the least understood areas of the type system by random contributors) so we remove all the evaluation methods forcing only the normalization routines to be used which will always be correct

Reply:
> `normalize_erasing_regions(c).to_valtree()`

So the idea is that `normalize_erasing_regions`  will invoke the interpreter if possible to make this a valtree? And `to_valtree()` is basically "check that this is already a valtree, fail otherwise"?

> evaluation is a subset of type system normalization

Indeed. It's its own huge system that gets invoked by the other huge system. Now we have to figure out how to make them work together best. :)

It can also be the other way around though: if I have a constant and want to evaluate it (a common thing one might want to do in the compiler), maybe I first need to do some type system normalization. Then type system normalization becomes part of const evaluation.

I feel fairly strongly that this operation should be called `eval` as that's what it does. I understand it needs to interact with type system normalization. But that could easily be done by having `mir::Const::eval` call type system normalization, no? The entire action that takes a `mir::Const` and returns a `mir::ConstValue` surely can be put into a helper function, even if it needs to do careful type system interaction things. That's exactly what `mir::Const::eval` is supposed to do (and dito for `ty::Const::eval`).

Reply:
normalize_erasing_regions *already* invokes the interpreter. type system normalization machinery is *already* capable of taking care of fully normalizing and evaluating constants in the type system.

the *only* thing that we are changing is that you must now use that api when you want to normalize a *type system constant* rather than using the `eval_x` methods.

> It can also be the other way around though

It cant because this is *really error prone* normalization of aliases has to return goals to be proven and the existing API for normalization of things in the type system through `ocx/fcx.normalize` exists to no longer have the kind of error prone API that makign `eval_x` also do alias normalization would introduce.

I mentioned this in the other comment thread:
> if we were to try and update the `eval_x` methods to be the kind of normalization that the type system requires then we would have to return the nested goals of the normalization to the caller which in the past has caused soundness bugs.

Reply:
> if I have a constant and want to evaluate it (a common thing one might want to do in the compiler),

right now perhaps yes, in the future you will *never* want to "evaluate it" you will want to actually normalize it by going through the type system as otherwise you will get incorrect results. So, it really won't be that common of a thing to want to do ^^

The circumstances where it would be correct to just evaluate the constant without doing any kind of type system logic would pretty much be *only* the implementation of normalization of consts in the type system.

Reply:
> normalize_erasing_regions already invokes the interpreter.

What are the entry points for that? I thought it'd all go through `ty::Const::eval`.

> It cant because this is really error prone normalization of aliases has to return goals to be proven and the existing API for normalization of things in the type system through ocx/fcx.normalize exists to no longer have the kind of error prone API that makign eval_x also do alias normalization would introduce.

You said similar words in the other thread and I didn't understand them there either.^^  `noramlize_erasing_regions` exists and doesn't return any goals. And there are other, similar APIs, like `try_instantiate_mir_and_normalize_erasing_regions` that the interpreter uses. These aren't all unsound, are they?

I get the impression there are certain *parts of the compiler* where calling the "old" `eval` is a problem, and those are the parts you are most concerned about and are now designing the API around. But there are other parts of the compiler where it's not a problem at all and we shouldn't disregard their existence.

>  in the future you will never want to "evaluate it" you will want to actually normalize it by going through the type system as otherwise you will get incorrect results. So, it really won't be that common of a thing to want to do ^^

If that's what it means to "evaluate a constant", then that's what `Const::eval` should do.

Reply:
> What are the entry points for that? I thought it'd all go through ty::Const::eval.

Probably by directly calling `tcx.const_eval_resolve_for_typeck` in the normalization folders. Right now we just call `normalize` because we have way too many normalization folders right now :melting_face:

>  get the impression there are certain parts of the compiler where calling the "old" eval is a problem, and those are the parts you are most concerned about and are now designing the API around. But there are other parts of the compiler where it's not a problem at all and we shouldn't disregard their existence.

So ideally users should not need to manually call `eval` ever, for type-system consts **or** mir consts. We currently maintain invariants in the type system that types are fully normalized (or as normalized as they can get, for still-not-fully-substituted/polymorphic bodies) when you're substituting and instantiating a MIR body, and consts should be fully normalized too. Unfortunately, this doesn't happen *today* for `mir::Const` because MIR consts are ignored by the type-system folder that is responsible for normalization, but I'm considering whether that should be changed.

As for normalizing `ty::Const` specifically in MIR transformations, codegen, etc., this shouldn't be necessary anymore because we *do* eagerly normalize. So I understand that this may introduce a perceived inconsistency because if you're used to working with `mir::Const` and needing to call `eval` or one of the helpers like `eval_to_target_usize`, you might feel like it's missing. Let me know if and where I should leave a comment to clarify that for MIR users.

> If that's what it means to "evaluate a constant", then that's what Const::eval should do: run the combined mutually recursive machinery of type system normalization and const evaluation.

Well, part of the point is that users shouldn't need to call `eval`. The only inconsistency is if you're conjuring up a `ty::ConstKind::Unevaluated` yourself -- then you'd (after this PR) need to call `normalize_erasing_regions` on it. Whether that's the best API or if we should maintain `eval` as a clearer way of normalizing a "fresh" unevaluated const (or, imo, probably reintroduce *after* all of this) is I guess up for debate.